art-decisions/proposals/2022-11-20-arcaflow-signals.md

```diff
@@ -0,0 +1,211 @@
+delay the evaluation of that expression until the signal channel is closed and all messages of a kind are received.
+
+For example:
+
+```yaml
+steps:
+  foo:
+    ...
+  bar:
+    input:
+      baz: !expr collect($.steps.foo.signals.something)
+```
+
+### Outputs
+
+The current normal output results will also be transformed into signals, which depend on the completion of a plugin.
+However, a plugin is now allowed to not declare an output and work with signals instead.
+
+The workflow outputs will work as before. They act as a signal channel that will only ever receive one signal and then
+be closed.
+
+Workflow outputs can reference outputs in the same way and will receive the single signal that is sent in the event that
+the plugin outputs that particular output.
+
+If workflow outputs reference non-output signals, they should use the `collect()` function or similar to make sure they
+capture all signals in a particular channel if that is desired.
+
+### Prematurely stopping plugin execution
+
+With the help of a signal, a workflow author can stop the execution of a plugin prematurely. The plugin will receive a
+notification of the signal and may decide to end correctly or crash.
+
+The workflow author may declare this as follows (syntax may change):
+
+```yaml
+steps:
+  uperf_client:
+    ...
+  uperf_server:
+    stop_if:
+      - !expr $.steps.uperf_client.state.finished
+```
+
+If such a situation happens, the plugin will receive a `SIGTERM`. The plugin is allowed to emit an output or additional signals, but must then terminate in an orderly fashion. Should the plugin not terminate within a timeout (default: 30 seconds), the plugin will be stopped with a `SIGKILL`. Python SDK users will get appropriate hooks to interrupt the execution.
+
+### Step states
+
+First, we introduce the following states for a step:
+
+1. **Deployed:** The plugin has been deployed, but no execution has started yet. This is useful for the deployer to emit information, such as the IP Address.
+2. **Running:** The plugin execution has started and the plugin can now send and receive signals.
+3. **Crashed:** The plugin execution has ended unexpectedly. This will cause the workflow to fail.
+4. **Finished:** The plugin execution has finished normally.
+
+### Effects on the dependency tree
+
+We will, however, introduce the following new nodes for each step:
+
+1. `state:deployed`
+2. `state:running`
+3. `state:crashed`
+4. `state:finished`
+
+In addition, we will also introduce nodes for each of the signals the plugin emits, prefixed with `signal:`. The
+outputs will be prefixed with `output:`.
+
+In order to ensure that a workflow can be executed, signals constitute a dependency on the step in the dependency tree.
+This means, that two steps cannot depend on each other's signals, nor can three or more steps form a dependency circle.
+This constitutes a limitation, because no two steps can form a constant back and forth communication via signals.
```

Comment:
I am not following why we want simplicity after introducing the complexity of signals, and why can't we warn the user when it is not safe instead of "knowing better then they" and preventing it?